src/refactor/ftp.py

```diff
@@ -0,0 +1,238 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+'''
+'''
+ * @file ftp.py
+ * @author Robert Taylor
+ * @date 2022-09-21
+'''
+
+from groundStation import GroundStation
+from options import optionsFactory
+from system import GroundNodes, services
+import os
+from random import randint
+from receiveParser import ReceiveParser
+from system import services
+
+#TODO: not a fan of the level of object orientation here. FTP is responsible for too much
+#TODO: The naming conventions here are... interesting
+
+class ftp(GroundStation):
+    def __init__(self, opts):
+        super(ftp, self).__init__(opts)
+        self.services = services
+        self.receiveParse = ReceiveParser()
+        self.operation = ''
+        self.infile = ''
+        self.outfile = ''
+        if opts.get and opts.post:
+            raise ValueError("Cannot post and get a file at the same time")
+        if not opts.get and not opts.post:
+            raise ValueError("Must get or post a file")
+        
+        if opts.get:
+            self.operation = "get"
+            self.infile = opts.get
+        elif opts.post:
+            self.operation = "post"
+            self.infile = opts.post
+
+        if opts.outfile:
+            self.outfile = opts.outfile
+        else:
+            self.outfile = self.infile
+
+        self.blocksize = 512 # There's really no reason to change it..
+
+        self.burst_size = opts.burst_size
+        if (opts.sband):
+            self.dest_addr = GroundNodes.SBAND.value
+        else:
+            self.dest_addr = GroundNodes.GND.value
+
+        self.destPort = self.services.get("FTP_COMMAND").get("port")
+
+    def run(self):
+        if self.operation == "get":
+            self._do_get_request()
+        elif self.operation == "post":
+            self._do_post_request()
+        else:
+            raise ValueError("Not doing get or post request? What is my purpose?")
+
+    def setOperation(self, op):
+        if op == "get":
+            self.operation = op
+        elif op == "post":
+            self.operation = op
+        else:
+            raise ValueError("Invalid operation {}".format(op))
+
+    def setOutfile(self, file):
+        self.outfile = file
+
+    def setInfile(self, file):
+        self.infile = file
+
+    def _transaction(self, data):
+        try:
+            self.networkManager.send(self.satelliteAddr, self.destPort, data)
+            response = self.networkManager.receive(self.satelliteAddr, self.destPort, 10000)
+            return self.receiveParse.parseReturnValue(self.destPort, response)
+        except Exception as e:
+            print(e)
+            exit(1)
+
+    def _get_data_upload_packet(self, req_id, data, count):
+        subservice = self.services.get('FTP_COMMAND').get('subservice').get('FTP_UPLOAD_PACKET').get('subPort')
+        out = bytearray()
```

Comment:
Not necessary, but it's almost like the packets could be generated by parsing a JSON command descriptor. I'm not sure it would save typing or enhance clarity, unless you need to add another 10 commands

Reply:
I'd like it if this code worked the same as the updater, where it uses the real parser then modifies the packets, but I wanted to get the basics done